GitUI/CommandsDialogs/FormBrowse.cs

```diff
@@ -2818,78 +2822,105 @@ private void SubmoduleStatusProvider_StatusUpdated(object sender, SubmoduleStatu
             mi.Click += UpdateAllSubmodulesToolStripMenuItemClick;
-            newItems.Add((mi, null));
+            newItems.Add(mi);
 
             if (result.CurrentSubmoduleName is not null)
             {
                 var item = new ToolStripMenuItem(_updateCurrentSubmodule.Text) { Tag = result.CurrentSubmoduleName };
                 item.Click += UpdateSubmoduleToolStripMenuItemClick;
-                newItems.Add((item, null));
+                newItems.Add(item);
             }
 
             // Using AddRange is critical: if you used Add to add menu items one at a
             // time, performance would be extremely slow with many submodules (> 100).
-            toolStripButtonLevelUp.DropDownItems.AddRange(newItems.Select(e => e.item).ToArray());
+            toolStripButtonLevelUp.DropDownItems.AddRange(newItems.ToArray());
 
-            // Load details sequentially to not spawn too many background threads
-            // then refresh all items at once with a single switch to the main thread
-            ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            return newItems;
+        }
+
+        private async Task UpdateSubmoduleMenuStatus(SubmoduleInfoResult result, CancellationToken cancelToken)
+        {
+            if (_currentSubmoduleMenuItems is null)
             {
-                var loadDetails = newItems.Select(e => e.loadDetails).WhereNotNull();
-                var refreshActions = new List<Action>();
-                foreach (var loadFunc in loadDetails)
+                return;
+            }
+
+            await this.SwitchToMainThreadAsync(cancelToken);
+
+            Validates.NotNull(result.TopProject);
+            var infos = result.AllSubmodules.ToDictionary(info => info.Path, info => info);
+            infos[result.TopProject.Path] = result.TopProject;
+            foreach (var item in _currentSubmoduleMenuItems)
+            {
+                var path = (string)item.Tag;
+                if (GitExtUtils.Strings.IsNullOrWhiteSpace(path))
                 {
-                    cancelToken.ThrowIfCancellationRequested();
-                    var action = await loadFunc();
-                    refreshActions.Add(action);
+                    // not a submodule
+                    continue;
                 }
 
-                await this.SwitchToMainThreadAsync(cancelToken);
-                foreach (var refreshAction in refreshActions)
+                if (infos.ContainsKey(path))
                 {
-                    refreshAction();
+                    UpdateSubmoduleMenuItemStatus(item, infos[path]);
+                    infos.Remove(path);
                 }
-            }).FileAndForget();
+                else
+                {
+                    // structure no longer matching
+                    Debug.Assert(true, $"Status info with {1 + result.AllSubmodules.Count} records do not match current nodes ({_currentSubmoduleMenuItems.Count})");
```

Comment:
Would it log a message?

Reply:
This should never happen but is not a catastrophe if it does, so I believe this is OK to not log

Reply:
`Debug.Fail(...)` or `Debug.Assert(false, ...)`?

Reply:
I changed to Debug.Fail